clickhouse_test.go

```diff
@@ -922,6 +922,105 @@ func Test_Temporary_Table(t *testing.T) {
 	}
 }
 
+func Test_Select_External_Tables(t *testing.T) {
+	const (
+		ddl = `
+			CREATE TABLE clickhouse_test_select_external_tables (
+				string1  String,
+				string2  String
+			) Engine=Memory
+		`
+		dml = `
+			INSERT INTO clickhouse_test_select_external_tables (
+				string1,
+				string2
+			) VALUES (
+				?,
+				?
+			)
+		`
+		query      = `SELECT COUNT(*) FROM clickhouse_test_select_external_tables WHERE string1 IN ? AND string2 IN ? AND string1 NOT IN ?`
+		queryNamed = `SELECT COUNT(*) FROM clickhouse_test_select_external_tables WHERE string1 IN @e1 AND string2 IN @e2 AND string1 NOT IN @e3`
```

Comment:
do we have some benchmarks on how this compares.. `external tables` vs large `IN clause` ?

Reply:
We can add them
Although from previous tests we know the db breaks down completely with large IN clauses

Reply:
also slightly related to this, seems we should have an index whenever possible on `domain` when using a large IN clause,  https://github.com/ClickHouse/ClickHouse/pull/9261#issuecomment-592759264

Reply:
oh right, totally forgot about those max ast issues :)

Reply:
```
package benchmarking

import (
	"database/sql"
	"database/sql/driver"
	"encoding/csv"
	"fmt"
	"github.com/alvarotuso/clickhouse-go"
	"github.com/alvarotuso/clickhouse-go/lib/column"
	"io"
	"log"
	"os"
	"strings"
	"testing"
)

const baseQuery = `
	SELECT count(*) FROM company_attributes.first_party_company_attributes
	WHERE advertisable_eid = '5L5IV3X4ZNCUZFMLN5KKOD' AND source = 'salesforce'
`

const queryWithExternal = baseQuery + ` 
	AND domain IN ?
`

const queryWithPlainValuesFormat = baseQuery + ` 
	AND domain IN (%s)
`

const clickHouseHost = "rollworks-clickhouse-staging.internal.adroll.com"
const domainsFile = "/Users/alvarotuso/accounts_external_data.csv"

func getDB(parameters string) *sql.DB {
	host := fmt.Sprintf("clickhouse://%s:9000", clickHouseHost)
	if parameters != "" {
		host += "?" + parameters
	}
	db, err := sql.Open("clickhouse", host)
	if err != nil {
		log.Fatal(err)
	}
	return db
}

func GetDomains() []string {
	domains := make([]string, 0)
	csvFile, err := os.Open(domainsFile)
	if err != nil {
		log.Fatal("Couldn't open the csv file", err)
	}
	r := csv.NewReader(csvFile)
	for {
		// Read each record from csv
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		domains = append(domains, record[0])
	}
	return domains
}

func RunWithExternal() {
	db := getDB("")
	domains := GetDomains()
	col, _ := column.Factory("domain", "String", nil)
	externalTable := clickhouse.ExternalTable{
		Name: "domains",
		Values: make([][]driver.Value, 0),
		Columns: []column.Column{col},
	}
	for _, domain := range domains {
		externalTable.Values = append(externalTable.Values, []driver.Value{domain})
	}

	result, err := db.Query(queryWithExternal, externalTable)
	if err != nil {
		log.Fatal(err)
	}
	var count int
	result.Next()
	err = result.Scan(&count)
	if err != nil {
		log.Fatal(err)
	}
}

func RunWithValues() {
	domains := GetDomains()
	domainPlaceholders := make([]string, len(domains))
	args := make([]interface{}, len(domains))
	for idx, domain := range domains {
		domainPlaceholders[idx] = "?"
		args[idx] = domain
	}
	query := fmt.Sprintf(queryWithPlainValuesFormat, strings.Join(domainPlaceholders, ","))
	db := getDB(fmt.Sprintf("max_query_size=%d", len(query)*10))
	result, err := db.Query(query, args...)
	if err != nil {
		log.Fatal(err)
	}
	var count int
	result.Next()
	err = result.Scan(&count)
	if err != nil {
		log.Fatal(err)
	}
}

func BenchmarkWithExternal(b *testing.B) {
	for n := 0; n < b.N; n++ {
		RunWithExternal()
	}
}

func BenchmarkWithValues(b *testing.B) {
	for n := 0; n < b.N; n++ {
		RunWithValues()
	}
}
```
This benchmark completes in 7s for external data, and doesn't seem to complete at all (or at least takes more than a few minutes)

Reply:
@kushagra391 strangely I didn't need to increase the ast limit this time, but the query limit needed to be increased for this benchmark. I wasn't able to get it to finish though 😅